### PR TITLE
Custom directives

### PR DESCRIPTION
**Pull request by no1seman**
Add availability to add custom directives to GraphQL schema

### CODE REVIEW

File: graphql/execute.lua [hunk 1]
```diff
@@ -278,14 +278,48 @@ local function getFieldEntry(objectType, object, fields, context)
       if argument and argument.value then
           positions[pos] = {
               name=argument.name.value,
-              value=arguments[argument.name.value]
+              value=arguments[argument.name.value],
```
Comment (olegrok) [inline graphql/execute.lua:281]: This change is not related to your patch.

File: test/unit/graphql_test.lua [hunk 2]
```diff
@@ -1057,3 +1058,34 @@ function g.test_boolean_coerce()
...
+end
+
+function g.test_util_filter()
+    local res = util.filter({ { name = 'a' }, { name = 'b' }, }, function(v) return v.name == 'a' end)
+    t.assert_items_equals(res, {{name = 'a'}})
+end
+
+function g.test_util_values()
+    local res = util.values({ a = { name = 'a' }, b = { name = 'b' }, })
+    t.assert_items_equals(res, {{name = "a"}, {name = "b"}})
```
Comment (olegrok) [inline test/unit/graphql_test.lua]: Here you use single and double quotes at the same time. Please be consistent.

File: graphql/execute.lua [hunk 3]
```diff
@@ -278,14 +278,48 @@ local function getFieldEntry(objectType, object, fields, context)
...
+    util.map_name(context.schema.directives or {}, function(directive, directive_name)
+      local supplied_directive = directiveMap[directive_name]
+      if supplied_directive ~= nil then
+        local directiveArgumentMap = {}
+        for _, argument in ipairs(supplied_directive.arguments or {}) do
+          directiveArgumentMap[argument.name.value] = argument
+        end
+
+        directives[directive_name] = util.map(directive.arguments or {}, function(argument, name)
+        local supplied = directiveArgumentMap[name] and directiveArgumentMap[name].value
```
Comment (olegrok) [inline graphql/execute.lua]: Seems your need to add indention for function body.

File: graphql/schema.lua [hunk 4]
```diff
@@ -99,6 +99,26 @@ end
...
+            directive.arguments[name] = argument
+          end
+
+          if type(argument.kind) == 'string' then
+            argument.kind = types.resolve(argument.kind, self.name)
+          end
+          -- END_HACK: resolve type names to real types
+
+          local argumentType = argument.__type and argument or argument.kind
+          assert(argumentType, 'Must supply type for argument "' .. name .. '" on "' .. directive.name .. '"')
```
Comment (olegrok) [inline graphql/schema.lua]: Assert with concatenation looks bad. It's performed even assert isn't triggered. Pease change to `if ... then error() end` if your really need it.

File: graphql/util.lua [hunk 5]
```diff
@@ -169,17 +179,17 @@ local function coerceValue(node, schemaType, variables, opts)
...
+--
+-- @param obj value to check
+--
+-- @tparam string obj_name name of the value to form an error
+--
+-- @tparam string type_1
+-- @tparam[opt] string type_2
+-- @tparam[opt] string type_3
+--
+-- @return nothing
```
Comment (olegrok) [inline graphql/util.lua]: Such changes should be moved into separate patch.

File: test/unit/graphql_test.lua [hunk 6]
```diff
@@ -1057,3 +1058,34 @@ function g.test_boolean_coerce()
...
+    t.assert_equals(util.cmpdeeply({ a = { a = 1 }}, { a = { a = 2 }}), false)
+    t.assert_equals(util.cmpdeeply({ a = 1 }, { a = 1, b = { b = 3 }}), false)
+    t.assert_equals(util.cmpdeeply({{{ a = tonumber('nan')}}}, {{{ a = tonumber('nan')}}}), true)
+    t.assert_equals(util.cmpdeeply('nan', {{{ a = tonumber('nan')}}}), false)
+end
+
+function g.test_util_check()
+    t.assert_equals(select(2, pcall(util.check, 'a', 'a', nil)), 'a must be a nil, got string')
+    t.assert_equals(select(2, pcall(util.check, 'a', 'a', nil, 1)), 'a must be a nil or a 1, got string')
+    t.assert_equals(select(2, pcall(util.check, 'a', 'a', 3, 2, 1)), 'a must be a 3 or a 2r a 1, got string')
```
Comment (olegrok) [inline test/unit/graphql_test.lua]: I guess it should be changed to assert_error

**Review by olegrok — COMMENTED** [review]
Thanks for your patch. Please add commit message with detailed description for your changes. Also see my comments below.

File: graphql/schema.lua (hunk 4 above)
Comment (no1seman) [reply graphql/schema.lua]: Are u shure? I've made this exactly the same way as in other places of this module. You have plans to refactor all the module?

Comment (olegrok) [reply graphql/schema.lua]: I'm completely sure. You can easily prove it using memprof.
I don't have plans to refactor this module. For now we could only write optimal code and probably fix other places one day.

File: graphql/execute.lua (hunk 1 above)
Comment (no1seman) [reply graphql/execute.lua:281]: Resolved

**Comment by DifferentialOrange** [comment]
I was finishing this PR as a part of this sprint tasks. You can see reworked version in #26 . Since this PR is based on branch of no1seman/graphql which I have no access to, I have opened a new PR.

**Comment by Totktonada** [comment]
Pushed with several fixups in PR #26.
